src/lang/base/TypeChecker.ml

```diff
@@ -539,13 +539,14 @@ module ScillaTypechecker
       | [] -> pure () in
     let env0 = TEnv.copy TEnv.mk in
     foldM lib_vars ~init:([], env0)
-      ~f:(fun (entry_acc, env_acc) (rn, body) ->
+      ~f:(fun (entry_acc, env_acc) (rn, topt, body) ->
           wrap_with_info
             (sprintf "Type error when checking recursion primitive %s:\n"
                (get_id rn), dummy_loc) @@
           let%bind ((_, (ar, _)) as typed_body) = type_expr env0 body in
+          let%bind _ = match topt with Some tannot -> assert_type_equiv tannot ar.tp | None -> pure () in
```

Comment:
As far as I remember `assert_type_equiv` checks whether the two types are identical up to alpha-renaming (renaming of type variables).

This means that if we have something like this:

```
let f : 'A => Uint128 -> Uint128 =
  tfun 'A =>
  fun (x : 'A) =>
  x
```

One could argue that this is perfectly fine because the user expects to be able to apply the function to a Uint128, and is indeed able to do so. On the other hand, the function is more general than that, and can be applied to values of other types as well, so the stated type is incorrect.

So it's not an incorrect implementation per se - we just need to agree on the design. Any comments on this @vaivaswatha or @anton-trunov?

Reply:
>the function is more general than that, and can be applied to values of other types as well, so the stated type is incorrect.
I would go with this. i.e., the type annotation has to exactly match the (type-checker) inferred type.

Reply:
> I would go with this. i.e., the type annotation has to exactly match the (type-checker) inferred type.

This does look like a reasonable choice. We can always weaken this restriction if there is a need.

I can't see right now a use case for a type like `forall. 'A -> Uint128 -> Uint128`. Perhaps we should warn the user that `'A` is not being used.